File: csharp/tests/Toolnexus.Tests/Spike/Agents.cs
// SPIKE — Level-1/2 UX sugar over the substrate, C# port of js/spike/agents.ts.
// One new noun: Agent. Everything compiles to the six verbs in AgentRuntime.cs.
//
//   var explore = new Agent("explore", new AgentSpec { Does = "...", Uses = [lookup] });
//   var coder   = new Agent("coder",   new AgentSpec { Does = "...", SoulFile = "AGENTS.md", Team = [explore] });
//   var r = await coder.RunAsync(rtOpts, "fix the bug");
//
//   var mia = Agents.AgentFromDir("~/mia");        // the directory IS the agent
//   var started = Agents.StartAgent(mia, rtOpts, everyMs: 30_000);

namespace Toolnexus.Tests.Spike;

public sealed class AgentSpec
{
    /// <summary>What the delegating model sees — the routing description.</summary>
    public string Does = "";
    /// <summary>The toolkit VIEW for this agent (scoping = the whole security model).</summary>
    public List<ITool>? Uses;
    /// <summary>Identity: inline text (soul file contents).</summary>
    public string? Soul;
    /// <summary>Path form of soul — read at build time.</summary>
    public string? SoulFile;
    /// <summary>Task-tool targets: listing agents here IS the subagent wiring.</summary>
    public List<Agent>? Team;
    public Budget? Budget;
    /// <summary>Model id; "inherit" = runtime default.</summary>
    public string? Model;
    /// <summary>§10 interpreter authority for this agent's subtree.</summary>
    public Func<Request, Task<Answer>>? WaitFor;
    public Func<Handle, Task>? OnSpawn;
    public Func<Handle, string, Task>? OnClose;
}

public sealed class Agent
{
    public readonly string Name;
    public readonly AgentSpec Spec;

    public Agent(string name, AgentSpec spec)
    {
        Name = name;
        Spec = spec;
    }

    /// <summary>Collect this agent + its whole team graph into a runtime registry.</summary>
    public Dictionary<string, AgentDef> Registry(Dictionary<string, AgentDef>? acc = null)
    {
        acc ??= new Dictionary<string, AgentDef>();
        if (acc.ContainsKey(Name)) return acc;
        var soul = Spec.Soul ?? "";
        if (Spec.SoulFile != null) soul = File.ReadAllText(Spec.SoulFile);
        acc[Name] = new AgentDef
        {
            Name = Name,
            Description = Spec.Does,
            SystemPrompt = soul,
            Model = Spec.Model ?? "inherit",
            Tools = Spec.Uses,
            Budget = Spec.Budget,
            WaitFor = Spec.WaitFor,
            OnSpawn = Spec.OnSpawn,
            OnClose = Spec.OnClose,
            // team scoping: task targets = ONLY this agent's team (not the whole registry)
            TaskTargets = (Spec.Team ?? new List<Agent>()).Select(a => a.Name).ToList(),
        };
        foreach (var t in Spec.Team ?? new List<Agent>()) t.Registry(acc);
        return acc;
    }

    /// <summary>Level 1: one-shot — build a runtime, run to completion, tear down.</summary>
    public async Task<TaskResult> RunAsync(RuntimeOptions rtOpts, string prompt)
    {
        var rt = new AgentRuntime(rtOpts.CloneWithRegistry(Registry()));
        var spawned = rt.Spawn(rt.Root, Name);
        if (spawned.Error != null) return new TaskResult(spawned.Error, true, "done", 0, 0);
        var wait = rt.WaitAsync(spawned.Handle!);
        rt.Wake(spawned.Handle!, prompt);
        var r = await wait.ConfigureAwait(false);
        if (r.Status != "pending") await rt.CloseAsync(rt.Root).ConfigureAwait(false);
        return r;
    }

    /// <summary>Bridge: an Agent IS a Tool — drop it into the OLD API's extra tools.</summary>
    public ITool AsTool(RuntimeOptions rtOpts)
        => NativeTool.OfAsync(Name, Spec.Does,
            new Dictionary<string, object?>
            {
                ["type"] = "object",
                ["properties"] = new Dictionary<string, object?>
                {
                    ["prompt"] = new Dictionary<string, object?> { ["type"] = "string" },
                },
                ["required"] = new List<object?> { "prompt" },
            },
            async (args, _) =>
            {
                var prompt = args.TryGetValue("prompt", out var p) ? p?.ToString() ?? "" : "";
                var r = await RunAsync(rtOpts, prompt).ConfigureAwait(false);
                return new ToolResult(r.Text, r.IsError, new Dictionary<string, object?>
                {
                    ["agent"] = Name, ["turns"] = r.Turns, ["totalTokens"] = r.TotalTokens,
                });
            });
}

public sealed class StartedAgent
{
    public required Handle Handle;
    public required AgentRuntime Runtime;
    internal CancellationTokenSource? TimerCts;
    internal Task? TimerTask;

    public async Task StopAsync()
    {
        TimerCts?.Cancel();
        if (TimerTask != null)
        {
            try { await TimerTask.ConfigureAwait(false); }
            catch (OperationCanceledException) { }
        }
        await Runtime.CloseAsync(Runtime.Root).ConfigureAwait(false);
    }
}

public static class Agents
{
    /// <summary>Level 2 bootstrap discovery order (openclaw order; all optional).</summary>
    public static readonly string[] BootstrapOrder =
    {
        "AGENTS.md", "SOUL.md", "IDENTITY.md", "USER.md", "TOOLS.md", "HEARTBEAT.md", "MEMORY.md",
    };

    public const long MaxBootstrapFileBytes = 2 * 1024 * 1024;

    /// <summary>The HEARTBEAT_OK sentinel — a silent wake produces no outbound message.</summary>
    public const string HeartbeatOk = "HEARTBEAT_OK";

    /// <summary>
    /// Level 2: the directory IS the agent. Discovers the bootstrap files (2MB cap each) and
    /// injects them as named "## file" sections at session start (cache doctrine: read memory in).
    /// </summary>
    public static Agent AgentFromDir(string dir, AgentSpec? opts = null, string? name = null)
    {
        var sections = new List<string>();
        foreach (var f in BootstrapOrder)
        {
            var p = Path.Combine(dir, f);
            if (!File.Exists(p)) continue;
            var size = new FileInfo(p).Length;
            var body = File.ReadAllText(p);
            if (size > MaxBootstrapFileBytes)
                body = body[..(int)Math.Min(body.Length, MaxBootstrapFileBytes)] + "\n[truncated: file exceeds bootstrap cap]";
            sections.Add($"## {f}\n\n{body.Trim()}");
        }
        var spec = opts ?? new AgentSpec();
        if (string.IsNullOrEmpty(spec.Does)) spec.Does = $"persona agent from {dir}";
        spec.Soul = string.Join("\n\n", sections);
        return new Agent(name ?? Path.GetFileName(dir.TrimEnd(Path.DirectorySeparatorChar)), spec);
    }

    /// <summary>
    /// Level 2: start a long-lived persona — the heartbeat posts a tick to its own inbox
    /// (unsolicited rail; ticks coalesce) and wakes it; HEARTBEAT_OK results stay silent.
    /// </summary>
    public static StartedAgent StartAgent(Agent a, RuntimeOptions rtOpts, int? everyMs = null, Action<string>? onReport = null)
    {
        var rt = new AgentRuntime(rtOpts.CloneWithRegistry(a.Registry()));
        var spawned = rt.Spawn(rt.Root, a.Name);
        if (spawned.Error != null) throw new InvalidOperationException(spawned.Error);
        var handle = spawned.Handle!;
        var started = new StartedAgent { Handle = handle, Runtime = rt };
        if (everyMs is int ms)
        {
            var cts = new CancellationTokenSource();
            started.TimerCts = cts;
            started.TimerTask = Task.Run(async () =>
            {
                using var timer = new PeriodicTimer(TimeSpan.FromMilliseconds(ms));
                while (await timer.WaitForNextTickAsync(cts.Token).ConfigureAwait(false))
                {
                    rt.Post(handle, new InboxItem("clock", "timer", "tick"));
                    if (handle.State == "idle")
                    {
                        _ = rt.RunTurnAsync(handle,
                                "Heartbeat. Read your HEARTBEAT.md section and follow it. If nothing needs attention, reply HEARTBEAT_OK.")
                            .ContinueWith(t =>
                            {
                                if (!t.IsCompletedSuccessfully) return;
                                var r = t.Result;
                                if (!r.IsError && !r.Text.Contains(HeartbeatOk)) onReport?.Invoke(r.Text);
                            }, TaskScheduler.Default);
                    }
                }
            });
        }
        return started;
    }
}

File: java/src/main/java/io/github/muthuishere/toolnexus/agents/Agents.java
package io.github.muthuishere.toolnexus.agents;

import io.github.muthuishere.toolnexus.Answer;
import io.github.muthuishere.toolnexus.Request;
import io.github.muthuishere.toolnexus.Tool;
import io.github.muthuishere.toolnexus.ToolContext;
import io.github.muthuishere.toolnexus.ToolResult;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * The Level-1 surface (SPEC §7D "Level-1 surface"): {@code agent(name, spec)} →
 * {@link Agent#run(RuntimeOptions, String)} (one-shot) and {@link Agent#asTool(RuntimeOptions)}
 * (the bridge into the classic API's {@code extraTools} — the axiom's other direction: an Agent
 * IS a Tool). Everything compiles down to the six verbs of {@link AgentRuntime}.
 *
 * <p>Team scoping is the wiring model: listing agents in {@link AgentSpec#team} IS the subagent
 * graph. A runtime registry is the transitive closure of the entry agent's team graph —
 * unreachable agents are not present. Children get no {@code task} reach unless their own spec
 * declares a team (recursion is opt-in, never default).
 */
public final class Agents {

    private Agents() {}

    /** The heartbeat sentinel — a wake whose result contains it produces no outbound report. */
    public static final String HEARTBEAT_OK = "HEARTBEAT_OK";

    /** Bootstrap discovery order for {@link #agentFromDir}, injected as {@code ## <file>} sections. */
    public static final List<String> BOOTSTRAP_ORDER = List.of(
            "AGENTS.md", "SOUL.md", "IDENTITY.md", "USER.md", "TOOLS.md", "HEARTBEAT.md", "MEMORY.md");
    static final long MAX_BOOTSTRAP_FILE_BYTES = 2L * 1024 * 1024;

    /** The declarative spec behind {@link #agent(String, AgentSpec)} — SPEC §7D field-for-field:
     * {@code does, uses(tools), soul|soulFile, team, budget, model("inherit"), waitFor, onSpawn,
     * onClose}. */
    public static final class AgentSpec {
        /** Optional explicit name ({@link #agentFromDir} override). */
        public String name;
        /** What the delegating model sees — the required routing description. */
        public String does;
        /** The toolkit VIEW for this agent (scoping is the whole security model). */
        public List<Tool> tools;
        /** Identity: inline soul text… */
        public String soul;
        /** …or a soul file path (AGENTS.md / SOUL.md), read at registry-build time. */
        public Path soulFile;
        /** {@code task}-tool targets: listing agents here IS the subagent wiring. */
        public List<Agent> team;
        public Budget budget;
        /** Model id; {@code null} ⇒ {@code "inherit"} (the runtime default). */
        public String model;
        /** §10 interpreter authority for this agent's subtree. */
        public Function<Request, Answer> waitFor;
        public Consumer<Handle> onSpawn;
        public BiConsumer<Handle, String> onClose;

        public AgentSpec name(String v) { this.name = v; return this; }
        public AgentSpec does(String v) { this.does = v; return this; }
        public AgentSpec tools(Tool... v) { this.tools = List.of(v); return this; }
        public AgentSpec soul(String v) { this.soul = v; return this; }
        public AgentSpec soulFile(Path v) { this.soulFile = v; return this; }
        public AgentSpec team(Agent... v) { this.team = List.of(v); return this; }
        public AgentSpec budget(Budget v) { this.budget = v; return this; }
        public AgentSpec model(String v) { this.model = v; return this; }
        public AgentSpec waitFor(Function<Request, Answer> v) { this.waitFor = v; return this; }
        public AgentSpec onSpawn(Consumer<Handle> v) { this.onSpawn = v; return this; }
        public AgentSpec onClose(BiConsumer<Handle, String> v) { this.onClose = v; return this; }
    }

    /** The one new noun. */
    public static Agent agent(String name, AgentSpec spec) {
        return new Agent(name, spec);
    }

    public static final class Agent {
        public final String name;
        public final AgentSpec spec;

        Agent(String name, AgentSpec spec) {
            this.name = name;
            this.spec = spec;
        }

        /** Collect this agent + its whole team graph into a runtime registry (transitive closure
         * of the entry agent's team graph — unreachable agents are absent). */
        public Map<String, AgentDef> registry(Map<String, AgentDef> acc) {
            if (acc.containsKey(name)) return acc;
            String soul = spec.soul != null ? spec.soul : "";
            if (spec.soulFile != null) {
                try {
                    soul = Files.readString(spec.soulFile);
                } catch (IOException e) {
                    throw new UncheckedIOException(e);
                }
            }
            AgentDef def = new AgentDef(name, spec.does, soul, spec.model != null ? spec.model : "inherit");
            def.tools = spec.tools;
            def.budget = spec.budget;
            def.waitFor = spec.waitFor;
            def.onSpawn = spec.onSpawn;
            def.onClose = spec.onClose;
            // Team scoping: task targets = ONLY this agent's team (empty = no delegation at all —
            // recursion is opt-in, never default).
            List<String> targets = new ArrayList<>();
            if (spec.team != null) for (Agent a : spec.team) targets.add(a.name);
            def.team = targets;
            acc.put(name, def);
            if (spec.team != null) for (Agent a : spec.team) a.registry(acc);
            return acc;
        }

        public Map<String, AgentDef> registry() {
            return registry(new LinkedHashMap<>());
        }

        /** One-shot: build a runtime scoped to this agent's team graph, run the prompt to
         * completion, tear down (unless the run parked {@code pending} — a durable pending keeps
         * the tree suspended for {@link AgentRuntime#resume}). */
        public TaskResult run(RuntimeOptions rtOpts, String prompt) {
            AgentRuntime rt = new AgentRuntime(rtOpts.copyWithRegistry(registry()));
            AgentRuntime.Spawn sp = rt.spawn(rt.root, name);
            if (sp.error() != null) return new TaskResult(sp.error(), true, "error", null, null, 0, 0);
            var fut = rt.futureResult(sp.handle());
            rt.wake(sp.handle(), prompt);
            TaskResult r = fut.join();
            if (!"pending".equals(r.status())) rt.close(rt.root);
            return r;
        }

        /** The bridge: an Agent IS a Tool — drop it into the classic API's {@code extraTools}.
         * The tool takes {@code {prompt}}, runs the agent's own loop in isolation, and returns
         * ONLY its final text + metadata {@code {agent, turns, totalTokens}}. */
        public Tool asTool(RuntimeOptions rtOpts) {
            Agent self = this;
            return new Tool() {
                @Override public String name() { return self.name; }
                @Override public String description() { return self.spec.does; }
                @Override public Map<String, Object> inputSchema() {
                    return Map.of("type", "object",
                            "properties", Map.of("prompt", Map.of("type", "string")),
                            "required", List.of("prompt"));
                }
                @Override public String source() { return "custom"; }
                @Override public ToolResult execute(Map<String, Object> args, ToolContext ctx) {
                    TaskResult r = self.run(rtOpts, String.valueOf(args.get("prompt")));
                    Map<String, Object> md = new LinkedHashMap<>();
                    md.put("agent", self.name);
                    md.put("turns", r.turns());
                    md.put("totalTokens", r.totalTokens());
                    return new ToolResult(r.text(), r.isError(), md);
                }
            };
        }
    }

    /**
     * Level 2: the directory IS the agent. Discovers (all optional, in {@link #BOOTSTRAP_ORDER}):
     * AGENTS.md, SOUL.md, IDENTITY.md, USER.md, TOOLS.md, HEARTBEAT.md, MEMORY.md — injected as
     * named {@code ## <file>} sections into the soul at session start.
     */
    public static Agent agentFromDir(Path dir, AgentSpec overrides) {
        List<String> sections = new ArrayList<>();
        for (String f : BOOTSTRAP_ORDER) {
            Path p = dir.resolve(f);
            if (Files.exists(p)) {
                try {
                    long size = Files.size(p);
                    String body = Files.readString(p);
                    if (size > MAX_BOOTSTRAP_FILE_BYTES) {
                        body = body.substring(0, (int) Math.min(body.length(), MAX_BOOTSTRAP_FILE_BYTES))
                                + "\n[truncated: file exceeds bootstrap cap]";
                    }
                    sections.add("## " + f + "\n\n" + body.strip());
                } catch (IOException e) {
                    throw new UncheckedIOException(e);
                }
            }
        }
        AgentSpec spec = overrides != null ? overrides : new AgentSpec();
        String name = spec.name != null ? spec.name : dir.getFileName().toString();
        if (spec.does == null) spec.does = "persona agent from " + dir;
        spec.soul = String.join("\n\n", sections);
        return agent(name, spec);
    }

    /** A long-lived persona started by {@link #startAgent}: its handle, its runtime, and the
     * heartbeat subscription. {@link #stop()} closes the whole tree. */
    public static final class StartedAgent {
        public final Handle handle;
        public final AgentRuntime rt;
        private final AutoCloseable heartbeat;

        StartedAgent(Handle handle, AgentRuntime rt, AutoCloseable heartbeat) {
            this.handle = handle;
            this.rt = rt;
            this.heartbeat = heartbeat;
        }

        public void stop() {
            if (heartbeat != null) {
                try {
                    heartbeat.close();
                } catch (Exception ignored) {
                    // a timer that fails to close must not block the close cascade
                }
            }
            rt.close(rt.root);
        }
    }

    /**
     * Level 2: start a long-lived persona. The heartbeat (driven by the runtime's injectable
     * {@link RuntimeClock} — virtual in fixtures) posts a tick to the agent's own inbox
     * (unsolicited rail; ticks coalesce) and wakes it; results containing {@link #HEARTBEAT_OK}
     * stay silent, anything else is reported.
     */
    public static StartedAgent startAgent(Agent a, RuntimeOptions rtOpts, Long everyMs,
                                          Consumer<String> onReport) {
        AgentRuntime rt = new AgentRuntime(rtOpts.copyWithRegistry(a.registry()));
        AgentRuntime.Spawn sp = rt.spawn(rt.root, a.name);
        if (sp.error() != null) throw new IllegalStateException(sp.error());
        Handle handle = sp.handle();
        AutoCloseable heartbeat = null;
        if (everyMs != null) {
            heartbeat = rt.clock().schedule(everyMs, () -> {
                rt.post(handle, new InboxItem("clock", "timer", "tick"));
                if (handle.state == Handle.State.IDLE) {
                    Thread.startVirtualThread(() -> {
                        TaskResult r = rt.runTurn(handle,
                                "Heartbeat. Read your HEARTBEAT.md section and follow it. "
                                        + "If nothing needs attention, reply HEARTBEAT_OK.");
                        if (!r.isError() && !r.text().contains(HEARTBEAT_OK) && onReport != null) {
                            onReport.accept(r.text());
                        }
                    });
                }
            });
        }
        return new StartedAgent(handle, rt, heartbeat);
    }
}
